finality-aleph/src/aggregation/multicast.rs

```diff
@@ -37,68 +32,26 @@ impl<H: Hash> Signable for SignableHash<H> {
     }
 }
 
-pub trait HasSignature<MK: MultiKeychain> {
-    fn signature(&self) -> MK::PartialMultisignature;
-}
-
-impl<H: Hash, MK: MultiKeychain> HasSignature<MK>
-    for UncheckedSigned<SignableHash<H>, MK::PartialMultisignature>
-{
-    fn signature(&self) -> MK::PartialMultisignature {
-        self.signature()
-    }
-}
-
-/// Anything that exposes the same interface as `aleph_bft::Multisigned` struct.
-pub trait Multisigned<'a, S: Signable, MK: MultiKeychain> {
-    type Result: HasSignature<MK> + Debug;
-    fn as_signable(&self) -> &S;
-    fn into_unchecked(self) -> Self::Result;
-}
-
-impl<'a, H: Hash, MK: MultiKeychain> Multisigned<'a, SignableHash<H>, MK>
-    for MultisignedStruct<'a, SignableHash<H>, MK>
-{
-    type Result = UncheckedSigned<SignableHash<H>, MK::PartialMultisignature>;
-    fn as_signable(&self) -> &SignableHash<H> {
-        self.as_signable()
-    }
-
-    fn into_unchecked(self) -> UncheckedSigned<SignableHash<H>, MK::PartialMultisignature> {
-        self.into_unchecked()
-    }
-}
-
 /// Anything that exposes the same interface as `aleph_bft::ReliableMulticast`.
 ///
 /// The trait defines an associated type: `Signed`. For `ReliableMulticast`, this is simply
 /// `aleph_bft::Multisigned` but the mocks are free to use the simplest matching implementation.
 #[async_trait::async_trait]
-pub trait Multicast<H: Hash, S: Signable>: Send + Sync {
-    type Signed;
-    async fn start_multicast(&mut self, signable: S);
-    fn get_multisigned(&self, signable: &S) -> Option<Self::Signed>;
-    async fn next_multisigned_hash(&mut self) -> Self::Signed;
+pub trait Multicast<H: Hash, PMS: PartialMultisignature>: Send + Sync {
+    async fn start_multicast(&mut self, signable: SignableHash<H>);
+    async fn next_signed_pair(&mut self) -> (H, PMS);
 }
 
 #[async_trait::async_trait]
-impl<'a, H: Hash, MK: MultiKeychain> Multicast<H, SignableHash<H>>
-    for ReliableMulticast<'a, SignableHash<H>, MK>
+impl<'a, H: Hash> Multicast<H, SignatureSet<Signature>>
+    for ReliableMulticast<'a, SignableHash<H>, KeyBox>
 {
-    type Signed = MultisignedStruct<'a, SignableHash<H>, MK>;
-
     async fn start_multicast(&mut self, hash: SignableHash<H>) {
         self.start_rmc(hash).await;
     }
 
-    fn get_multisigned(
-        &self,
-        hash: &SignableHash<H>,
-    ) -> Option<MultisignedStruct<'a, SignableHash<H>, MK>> {
-        self.get_multisigned(hash)
-    }
-
-    async fn next_multisigned_hash(&mut self) -> MultisignedStruct<'a, SignableHash<H>, MK> {
-        self.next_multisigned_hash().await
+    async fn next_signed_pair(&mut self) -> (H, SignatureSet<Signature>) {
+        let ms = self.next_multisigned_hash().await.into_unchecked();
+        (Signable::hash(ms.as_signable()), ms.signature())
```

Comment:
Uh, I think you want `.clone()` here rather than using `Signable::hash`. Unless I'm massively misinterpreting what is going on here.